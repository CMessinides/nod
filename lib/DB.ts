import { Pool } from "pg";
import { connectionString } from "../config/db.config";
import { dev } from "../config/server.config";

const pool = new Pool({ connectionString, ssl: !dev });

export default {
	query: pool.query
};
