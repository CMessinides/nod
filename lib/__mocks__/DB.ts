const DB = jest.genMockFromModule("../DB").default;

export function mockRows(rows = [], { once = true } = {}): void {
	const result = { rows, rowCount: rows.length };
	if (once) {
		DB.query.mockResolvedValueOnce(result);
	} else {
		DB.query.mockResolvedValue(result);
	}
}

DB.__mockRows = mockRows;
DB.__mockRows([], { once: false });

export default DB;
