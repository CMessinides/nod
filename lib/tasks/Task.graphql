type Task {
	id: ID!
	name: String!
	done: Boolean!
	createdAt: Date!
}
