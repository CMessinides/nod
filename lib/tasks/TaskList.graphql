type TaskList {
	id: ID!
	name: String!
	tasks: [Task]!
}
