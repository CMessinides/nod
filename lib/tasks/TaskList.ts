import { TaskListModel } from "./interfaces";
import DB from "../DB";
import Task from "./Task";

type TaskListRecord = {
	id: number;
	name: string;
};

type TaskListParams = Omit<TaskListModel, "tasks">;

export default class TaskList implements TaskListModel {
	id: number;
	name: string;

	static async findById(id: number) {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM task_lists WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return TaskList.fromRecord(rows[0] as TaskListRecord);
	}

	static fromRecord(record: TaskListRecord) {
		return new TaskList(record);
	}

	constructor(params: TaskListParams) {
		for (const key of Object.keys(params)) {
			this[key] = params[key];
		}
	}

	tasks() {
		return Task.getAllInList(this.id);
	}
}
