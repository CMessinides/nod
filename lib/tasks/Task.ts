import DB from "../DB";
import { TaskModel } from "./interfaces";
import TaskList from "./TaskList";
import LinkedList from "../LinkedList";

interface TaskRecord {
	id: number;
	name: string;
	done: boolean;
	created_at: Date;
	list_id: number;
	prev_task_id: number;
}

type TaskParams = Omit<TaskModel, "list"> & {
	listId: number;
};

export default class Task implements TaskModel {
	id: number;
	name: string;
	done: boolean;
	createdAt: Date;
	private listId: number;

	static async findById(id: number): Promise<Task> {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM tasks WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return Task.fromRecord(rows[0] as TaskRecord);
	}

	static async getAllInList(listId: number): Promise<Task[]> {
		const { rows: records }: { rows: TaskRecord[] } = await DB.query(
			"SELECT * FROM tasks WHERE list_id = $1",
			[listId]
		);

		return Array.from(LinkedList(records, "id", "prev_task_id")).map(
			Task.fromRecord
		);
	}

	static fromRecord({ id, name, done, created_at, list_id }: TaskRecord): Task {
		return new Task({
			id,
			name,
			done,
			createdAt: created_at,
			listId: list_id
		});
	}

	constructor(params: TaskParams) {
		for (const key of Object.keys(params)) {
			this[key] = params[key];
		}
	}

	list(): Promise<TaskList> {
		return TaskList.findById(this.listId);
	}
}
