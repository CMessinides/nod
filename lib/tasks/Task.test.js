import Task from "./Task";

// For mocking
import TaskList from "./TaskList";
import DB from "../DB";
import MockRecords from "./fixtures/TaskRecords";

jest.mock("../DB");
DB.query.mockResolvedValue({ rows: [], rowCount: 0 });

jest.mock("./TaskList");
TaskList.findById.mockResolvedValue(new TaskList());

describe("static methods", () => {
	describe("findById", () => {
		it("should return the task with the given ID", async () => {
			const record = MockRecords[0];
			const { id } = record;
			mockDBRows([record]);

			expect(Task.findById(id)).resolves.toMatchSnapshot();
		});

		it("should return null if the task does not exist", () => {
			// mock the database to return an empty array when queried for the task
			mockDBRows([]);

			expect(Task.findById(1)).resolves.toBe(null);
		});
	});

	describe("getAllInList", () => {
		it("should return the list's tasks in order", () => {
			const listId = 1;
			mockDBRows(MockRecords.filter(record => record.list_id === listId));

			expect(Task.getAllInList(listId)).resolves.toMatchSnapshot();
		});

		it("should return an empty array if the list does not exist", () => {
			// mock the database to return an empty array when queried for the list
			mockDBRows([]);

			expect(Task.getAllInList(1)).resolves.toHaveLength(0);
		});
	});
});

describe("instance methods", () => {
	describe("list", () => {
		it("should return the task's parent list", () => {
			const task = Task.fromRecord(MockRecords[0]);
			const list = new TaskList({
				id: task.listId
			});

			TaskList.findById.mockResolvedValueOnce(list);

			expect(task.list()).resolves.toBe(list);
		});
	});
});

function mockDBRows(rows) {
	DB.query.mockResolvedValueOnce({ rows, rowCount: rows.length });
}
