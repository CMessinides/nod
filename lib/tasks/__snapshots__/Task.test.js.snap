// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static methods findById should return the task with the given ID 1`] = `
Task {
  "createdAt": 1970-01-18T20:10:49.268Z,
  "done": false,
  "id": 1,
  "listId": 1,
  "name": "Soup - Campbells Chili",
}
`;

exports[`static methods getAllInList should return the list's tasks in order 1`] = `
Array [
  Task {
    "createdAt": 1970-01-18T20:10:49.268Z,
    "done": false,
    "id": 1,
    "listId": 1,
    "name": "Soup - Campbells Chili",
  },
  Task {
    "createdAt": 1970-01-19T00:31:47.828Z,
    "done": false,
    "id": 8,
    "listId": 1,
    "name": "Beans - Turtle, Black, Dry",
  },
  Task {
    "createdAt": 1970-01-19T01:21:40.186Z,
    "done": true,
    "id": 7,
    "listId": 1,
    "name": "Contreau",
  },
  Task {
    "createdAt": 1970-01-18T20:12:24.398Z,
    "done": false,
    "id": 2,
    "listId": 1,
    "name": "Lotus Rootlets - Canned",
  },
  Task {
    "createdAt": 1970-01-18T20:01:40.482Z,
    "done": true,
    "id": 4,
    "listId": 1,
    "name": "Turkey - Breast, Boneless Sk On",
  },
]
`;
